Lib/test/test_urlparse.py

```diff
@@ -579,6 +576,53 @@ def test_urljoins(self):
         # issue 23703: don't duplicate filename
         self.checkJoin('a', 'b', 'b')
 
+        # Test with empty (but defined) components.
+        self.checkJoin(RFC1808_BASE, '', 'http://a/b/c/d;p?q#f')
+        self.checkJoin(RFC1808_BASE, '#', 'http://a/b/c/d;p?q#', relroundtrip=False)
+        self.checkJoin(RFC1808_BASE, '#z', 'http://a/b/c/d;p?q#z')
+        self.checkJoin(RFC1808_BASE, '?', 'http://a/b/c/d;p?', relroundtrip=False)
+        self.checkJoin(RFC1808_BASE, '?#z', 'http://a/b/c/d;p?#z', relroundtrip=False)
+        self.checkJoin(RFC1808_BASE, '?y', 'http://a/b/c/d;p?y')
+        self.checkJoin(RFC1808_BASE, ';', 'http://a/b/c/;')
```

Comment:
Making a note, using parameter `;` removes the final portion of the path in the base url `/d` in this case.

Reply:
And this as per the rfc3986 requirement as given in the `5.4.  Reference Resolution Examples` and these tests handle them properly for the empty `?` cases correctly.